src/Services/GarbageCollectionService.php

```diff
@@ -25,22 +26,30 @@ public function collect(): void
      */
     private function collectExpiredSessions(): void
     {
-        $lifetime = LoginSession::config()->get('default_session_lifetime');
-        $sessions = LoginSession::get()->filter([
-            'LastAccessed:LessThan' => date('Y-m-d H:i:s', time() - $lifetime),
+        $maxAge = LoginSession::getMaxAge();
+        LoginSession::get()->filter([
+            'LastAccessed:LessThan' => $maxAge,
             'Persistent' => 0
-        ]);
-        $sessions->removeAll();
+        ])->removeAll();
     }
 
     /**
      * Collect all persistent LoginSession records where the associated RememberLoginHash has expired
      */
     private function collectImplicitlyExpiredSessions(): void
     {
+        $now = DBDatetime::now()->getTimestamp();
+        LoginSession::get()->filter([
+            'Persistent' => 1,
+            'LoginHash.ExpiryDate:LessThan' => date('Y-m-d H:i:s', $now),
+        ])->removeAll();
+
+        $maxAge = LoginSession::getMaxAge();
+        // If a persistent session has no login hash, use LastAccessed
         LoginSession::get()->filter([
+            'LastAccessed:LessThan' => $maxAge,
             'Persistent' => 1,
-            'LoginHash.ExpiryDate:LessThan' => date('Y-m-d H:i:s')
+            'LoginHash.ExpiryDate' => null,
         ])->removeAll();
```

Comment:
This isn't checking if there's no hash as the comment suggests - it's just checking if the hash shouldn't expire.
If the has shouldn't expire, why would we remove it? This change doesn't make sense to me...

Reply:
I've already included the changes from this PR
https://github.com/silverstripe/silverstripe-session-manager/pull/155
as well, it should be merged first

The issue is that LoginHash.ExpiryDate:LessThan will never match any LoginSession marked as persistent without an associated RememberMeLoginHash. Currently, it's possible (doesn't happen a lot, but I do have a couple of these in my production db), to have persistent LoginSession that don't have a RememberMeLoginHash anymore. When that happens, the garbage collector do not collect them anymore, and they stay permanently. It's not clear from the unit test if that's by design or not, but I really think it shouldn't be the case.